Kernel/Graphics/VirtIOGPU/VirtIOFrameBufferDevice.cpp

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (c) 2021, Sahan Fernando <sahan.h.fernando@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/Graphics/GraphicsManagement.h>
+#include <Kernel/Graphics/VirtIOGPU/VirtIOFrameBufferDevice.h>
+#include <LibC/sys/ioctl_numbers.h>
+
+#define DEVICE_EVENTS_READ 0x0
+#define DEVICE_EVENTS_CLEAR 0x4
+#define DEVICE_NUM_SCANOUTS 0x8
+
+namespace Kernel::Graphics {
+
+unsigned VirtIOFrameBufferDevice::next_device_id = 0;
+
+VirtIOFrameBufferDevice::VirtIOFrameBufferDevice(RefPtr<VirtIOGPU> virtio_gpu)
+    : BlockDevice(29, GraphicsManagement::the().allocate_minor_device_number())
+    , m_gpu(virtio_gpu)
+{
+    auto write_sink_page = MM.allocate_user_physical_page(MemoryManager::ShouldZeroFill::No).release_nonnull();
+    auto num_needed_pages = m_gpu->framebuffer_vm_object().page_count();
+    NonnullRefPtrVector<PhysicalPage> pages;
+    for (auto i = 0u; i < num_needed_pages; ++i) {
+        pages.append(write_sink_page);
+    }
+    m_framebuffer_sink_vmobject = adopt_ref(*new AnonymousVMObject(move(pages)));
+}
+
+VirtIOFrameBufferDevice::~VirtIOFrameBufferDevice()
+{
+}
+
+int VirtIOFrameBufferDevice::ioctl(FileDescription&, unsigned request, FlatPtr arg)
+{
+    REQUIRE_PROMISE(video);
+    switch (request) {
+    case FB_IOCTL_GET_SIZE_IN_BYTES: {
+        auto* out = (size_t*)arg;
+        size_t value = m_gpu->framebuffer_size_in_bytes();
+        if (!copy_to_user(out, &value))
+            return -EFAULT;
+        return 0;
+    }
+    case FB_IOCTL_GET_RESOLUTION:
+    case FB_IOCTL_SET_RESOLUTION: {
+        auto* user_resolution = (FBResolution*)arg;
+        FBResolution resolution;
+        resolution.pitch = m_gpu->framebuffer_pitch();
+        resolution.width = m_gpu->framebuffer_width();
+        resolution.height = m_gpu->framebuffer_height();
+        if (!copy_to_user(user_resolution, &resolution))
+            return -EFAULT;
+        return request == FB_IOCTL_GET_BUFFER ? 0 : -ENOTSUP;
+    }
+    case FB_IOCTL_FLUSH_BUFFER: {
+        FBRect user_dirty_rect;
+        if (!copy_from_user(&user_dirty_rect, (FBRect*)arg))
+            return -EFAULT;
+        VirtIOGPURect dirty_rect {
+            .x = user_dirty_rect.x,
+            .y = user_dirty_rect.y,
+            .width = user_dirty_rect.width,
+            .height = user_dirty_rect.height
+        };
+        m_gpu->flush_dirty_window(dirty_rect);
+        return 0;
+    }
+    default:
+        return -EINVAL;
+    };
+}
+
+KResultOr<Region*> VirtIOFrameBufferDevice::mmap(Process& process, FileDescription&, const Range& range, u64 offset, int prot, bool shared)
+{
+    REQUIRE_PROMISE(video);
+    if (!shared)
+        return ENODEV;
+    if (offset != 0)
+        return ENXIO;
+    if (range.size() != page_round_up(m_gpu->framebuffer_size_in_bytes()))
+        return EOVERFLOW;
+    // We only allow one process to map the region
+    if (m_userspace_mmap_region)
+        return ENOMEM;
```

Comment:
Is there a better error we can use here instead of ENOMEM? since thats not its usual meaning.

Reply:
There are more descriptive errors, but not any that are allowed to be returned by mmap according to posix. Maybe EMFILE?

Reply:
ENFILE is similar to what youre trying to convey, but not exactly, i would probably go for EINVAL, but i guess ENOMEM is fine :shrug:

Reply:
I think I'll stay with ENOMEM then :).